### PR TITLE
Documented C# when keyword

### LINKED ISSUE #1364: C# Language keywords: when
Added in C# 6 for Exception filters.
Updated in C# 7 for pattern matching switch statements.

The migrated C# Language reference does not contain this keyword.

In addition, the C# Language programming guide incorrectly states that exception filters are not supported in C#

### PR DESCRIPTION
**Pull request by rpetrusha**
# Documented C# `when` keyword

## Summary

Documented `when` contextual keyword used with `catch` and `case`. Insert a short (one or two sentence) summary here.

Fixes #1364 

## Details

I'll modify the TOC once its been updated to reflect existing open PRs.

## Suggested Reviewers

@BillWagner

### CODE REVIEW

File: docs/csharp/language-reference/keywords/when.md [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+## `when` in a `switch` statement
+
+Starting with 7, `case` labels no longer need be mutually exclusive, and the order in which `case` labels appear in a `switch` statement can determine which switch block executes. The `when` keyword can be used to specify a filter condition that causes its associated case label to be true only if the filter condition is also true. Its syntax is:
+
+```cs
+case (expr) where (when-condition):
+```
+where *expr* is a constant pattern or type pattern that is compared to the match expression, and *when-condition* is any Boolean expression. 
+
+The following example uses the `when` keyword to conditionally execute a switch section when a `Shape` object is `null` or has no area.
```
Comment (BillWagner) [inline docs/csharp/language-reference/keywords/when.md]: Same comment as the other PR: 'when' isn't necessary for a null check.

File: samples/snippets/csharp/language-reference/keywords/when/when.cs [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+                         sh, new Square(0) };
+      foreach (var shape in shapes)
+         ShowShapeInfo(shape);
+   }
+
+   private static void ShowShapeInfo(Shape sh)
+   {
+      switch (sh)
+      {
+         case Shape shape when shape == null:
```
Comment (BillWagner) [inline samples/snippets/csharp/language-reference/keywords/when/when.cs]: This will never PR. This is the same comment from the earlier PR.

**Review by BillWagner — CHANGES_REQUESTED** [review]
@rpetrusha Overall, this is great.

I have the same comments here as on the `switch`  PR with respect to the use of null on a when expression.

File: samples/snippets/csharp/language-reference/keywords/when/when.cs (hunk 2 above)
Comment (rpetrusha) [inline samples/snippets/csharp/language-reference/keywords/when/when.cs]: @BillWagner Thanks for catching this, Bill. I've modified the text and the example.

**Review by BillWagner — APPROVED** [review]

**Pull request merged**